lib/cartopy/mpl/gridliner.py

```diff
@@ -327,58 +338,257 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
 
         #################
         # Label drawing #
         #################
 
-        # Trim outside-area points from the label coords.
-        # Tickers may round *up* the desired range to something tidy, not
-        # all of which is necessarily visible.  We must be stricter with
-        # our texts, as they are drawn *without clipping*.
-        x_label_points = [x for x in x_ticks if x_lim[0] <= x <= x_lim[1]]
-        y_label_points = [y for y in y_ticks if y_lim[0] <= y <= y_lim[1]]
+        self.bottom_label_artists = []
+        self.top_label_artists = []
+        self.left_label_artists = []
+        self.right_label_artists = []
+        if not (self.left_labels or self.right_labels or
+                self.bottom_labels or self.top_labels):
+            return
+        self._assert_can_draw_ticks()
+
+        # Get the real map boundaries
+        map_boundary_vertices = self.axes.background_patch.get_path().vertices
+        map_boundary = sgeom.Polygon(map_boundary_vertices)
+
+        self._labels = []
+        for lonlat, lines, line_ticks, formatter, label_style in (
+                ('lon', lon_lines, lon_ticks,
+                 self.xformatter, self.xlabel_style),
+                ('lat', lat_lines, lat_ticks,
+                 self.yformatter, self.ylabel_style)):
+
+            formatter.set_locs(line_ticks)
+
+            for line, tick_value in zip(lines, line_ticks):
+
+                # Intersection of line with map boundary
+                line = np.array(line)
+                line = self.axes.projection.transform_points(
+                        crs, line[:, 0], line[:, 1])[:, :2]
+                infs = np.isinf(line)
+                if infs.any():
+                    if infs.all():
+                        continue
+                    line = line.compress(~infs.any(axis=1), axis=0)
+                line = sgeom.LineString(line)
+                if line.intersects(map_boundary):
+                    intersection = line.intersection(map_boundary)
+                    del line
+                    if intersection.is_empty:
+                        continue
+                    if isinstance(intersection, sgeom.MultiPoint):
+                        if len(intersection) < 2:
+                            continue
+                        tails = [[(pt.x, pt.y) for pt in intersection[:2]]]
+                        heads = [[(pt.x, pt.y)
+                                  for pt in intersection[-1:-3:-1]]]
+                    elif isinstance(intersection, (sgeom.LineString,
+                                                   sgeom.MultiLineString)):
+                        if isinstance(intersection, sgeom.LineString):
+                            intersection = [intersection]
+                        elif len(intersection) > 4:
+                            # gridline and map boundary are parallel
+                            # and they intersect themselve too musc
```

Comment:
should this say: `# and they intersect themselves too much`?

Reply:
Fixed